testing/dart/compositing_test.dart

```diff
@@ -51,4 +51,92 @@ void main() {
       throwsA(const TypeMatcher<AssertionError>()),
     );
   });
+
+  test('SceneBuilder accepts typed layers', () {
+    final SceneBuilder builder1 = SceneBuilder();
+    final OpacityEngineLayer opacity1 = builder1.pushOpacity(100);
+    expect(opacity1, isNotNull);
+    builder1.pop();
+    builder1.build();
+
+    final SceneBuilder builder2 = SceneBuilder();
+    final OpacityEngineLayer opacity2 = builder2.pushOpacity(200, oldLayer: opacity1);
+    expect(opacity2, isNotNull);
+    builder2.pop();
+    builder2.build();
+  });
+
+  void testNoSharing(_TestNoSharingFunction pushFunction) {
+    final SceneBuilder builder1 = SceneBuilder();
+    final EngineLayer layer = pushFunction(builder1, null);
+    builder1.pop();
+    builder1.build();
+
+    // Test: first push then attempt illegal addRetained
+    final SceneBuilder builder2 = SceneBuilder();
+    pushFunction(builder2, layer);
+    builder2.pop();
+    try {
+      builder2.addRetained(layer);
+      fail('Expected addRetained to throw AssertionError but it returned successully');
+    } on AssertionError catch (error) {
+      expect(error.toString(), contains('The layer is already being used'));
+    }
+    builder2.build();
+
+    // Test: first addRetained then attempt illegal push
```

Comment:
Maybe also test two pushes with the same oldLayer?

Reply:
Good idea! Done. Also added two `addRetained`s.